File: 166_fraction_to_decimal/fraction_to_decimal.go
/*
 * @Date: 2021-10-03 09:00:19
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-10-03 09:02:52
 */

package main

import "strconv"

func fractionToDecimal(numerator, denominator int) string {
	if numerator%denominator == 0 {
		return strconv.Itoa(numerator / denominator)
	}

	s := []byte{}
	if numerator < 0 != (denominator < 0) {
		s = append(s, '-')
	}

	abs := func(x int) int {
		if x < 0 {
			return -x
		}
		return x
	}
	// 整数部分
	numerator = abs(numerator)
	denominator = abs(denominator)
	integerPart := numerator / denominator
	s = append(s, strconv.Itoa(integerPart)...)
	s = append(s, '.')

	// 小数部分
	indexMap := map[int]int{}
	remainder := numerator % denominator
	for remainder != 0 && indexMap[remainder] == 0 {
		indexMap[remainder] = len(s)
		remainder *= 10
		s = append(s, '0'+byte(remainder/denominator))
		remainder %= denominator
	}
	if remainder > 0 { // 有循环节
		insertIndex := indexMap[remainder]
		s = append(s[:insertIndex], append([]byte{'('}, s[insertIndex:]...)...)
		s = append(s, ')')
	}

	return string(s)
}

func main() {
	assert := func(b bool) {
		if !b {
			panic("Not Passed")
		}
	}
	assert(fractionToDecimal(1, 2) == "0.5")
	assert(fractionToDecimal(2, 1) == "2")
	assert(fractionToDecimal(2, 3) == "0.(6)")
	assert(fractionToDecimal(4, 333) == "0.(012)")
	assert(fractionToDecimal(1, 5) == "0.2")
}

File: rust/src/166_fraction_to_decimal/fraction_to_decimal.rs
/*
 * @Date: 2021-10-03 09:05:53
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-10-03 09:09:35
 */

struct Solution;

use std::collections::HashMap;

impl Solution {
    pub fn fraction_to_decimal(numerator: i32, denominator: i32) -> String {
        if numerator == 0 {
            return "0".to_string();
        }
        let mut s = String::new();
        if (numerator < 0) ^ (denominator < 0) {
            s.push('-');
        }
        let (numerator, denominator) = ((numerator as i64).abs(), (denominator as i64).abs());
        let (integer, mut remainder) = (numerator / denominator, numerator % denominator);
        s.push_str(&integer.to_string());
        if remainder != 0 {
            s.push('.');
        }
        let mut map = HashMap::new();
        while remainder != 0 && !map.contains_key(&remainder) {
            map.insert(remainder, s.len());
            remainder *= 10;
            s.push_str(&(remainder / denominator).to_string());
            remainder %= denominator;
        }
        if let Some(pos) = map.get(&remainder) {
            s.insert(*pos, '(');
            s.push(')');
        }
        s
    }
}

fn main() {
    assert_eq!(Solution::fraction_to_decimal(1, 2), "0.5");
    assert_eq!(Solution::fraction_to_decimal(2, 1), "2");
    assert_eq!(Solution::fraction_to_decimal(2, 3), "0.(6)");
    assert_eq!(Solution::fraction_to_decimal(4, 333), "0.(012)");
    assert_eq!(Solution::fraction_to_decimal(1, 5), "0.2");
}
